juju/juju
# Feature/k8s service account

## Description of change

Add service account support for caas charm;

## QA steps

- prepare caas model;

```bash
$ juju bootstrap microk8s k1
$ juju add-model t1 microk8s
```

1. **Create a new Role/ClusterRole if rules are defined in role/clusterrole section**
- deploy k8s charm (for example, `mariadb-k8s`) with service account in podspec

```yaml
serviceAccount:
  name: build-robot
  automountServiceAccountToken: true
  capabilities:
    roleBinding:
      name: read-pods  # binding name
      type: ClusterRoleBinding  # RoleBinding/ClusterRoleBinding
    role:
      name: pod-reader
      type: ClusterRole  # Role/ClusterRole
      rules:
      - apiGroups: [""] # "" indicates the core API group
        resources: ["pods"]
        verbs: ["get", "watch", "list"]
```

- check resources;

```bash
$ kubectl get sa,roles,clusterroles,rolebindings,clusterrolebindings -n t1 --selector=juju-app=mariadb-k8s

NAME                         SECRETS   AGE
serviceaccount/build-robot   1         2m32s

NAME                                               AGE
clusterrole.rbac.authorization.k8s.io/pod-reader   29m

NAME                                                     AGE
clusterrolebinding.rbac.authorization.k8s.io/read-pods   2m32s
```

```bash
$ kubectl get statefulset mariadb-k8s -n t1 -o=jsonpath='{.spec.template.spec.serviceAccountName}'
build-robot
```

```bash
$ kubectl get sa build-robot -n t1 -o yaml
apiVersion: v1
automountServiceAccountToken: true
kind: ServiceAccount
metadata:
  creationTimestamp: "2019-07-05T08:44:10Z"
  labels:
    juju-app: mariadb-k8s
    juju-model: t1
  name: build-robot
  namespace: t1
  resourceVersion: "1099393"
  selfLink: /api/v1/namespaces/t1/serviceaccounts/build-robot
  uid: 1210cfec-00e2-44f5-973d-709948d1e77a
secrets:
- name: build-robot-token-s8px4
```

```bash
$ mkubectl get clusterrolebinding.rbac.authorization.k8s.io/read-pods -n t1 -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  creationTimestamp: "2019-07-04T09:43:32Z"
  labels:
    juju-app: mariadb-k8s
    juju-model: t1
  name: read-pods
  resourceVersion: "1074411"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/read-pods
  uid: 3089a75a-dd0f-4863-80e0-88834ad3f600
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: pod-reader
subjects:
- kind: ServiceAccount
  name: build-robot
  namespace: t1
```

2. **Reference to an existing Role/ClusterRole if NO rules are defined in role/clusterrole section**
- deploy k8s charm (for example, `mariadb-k8s`) with service account in podspec

```yaml
serviceAccount:
  name: build-robot
  automountServiceAccountToken: true
  capabilities:
    roleBinding:
      name: read-pods  # binding name
      type: RoleBinding
    role:
      name: nginx-ingress-microk8s-clusterrole  # this is an existing ClusterRole
      type: ClusterRole
      # if NO rules defined here, we think you are asking to ref to an existing ClusterRole named nginx-ingress-microk8s-clusterrole
```

- check resources;

```bash
$ kubectl get sa,roles,clusterroles,rolebindings,clusterrolebindings -o wide -n t1
NAME                         SECRETS   AGE
serviceaccount/build-robot   1         18s

NAME                                                                       AGE
clusterrole.rbac.authorization.k8s.io/nginx-ingress-microk8s-clusterrole   10d

NAME                                              AGE   ROLE                                             USERS   GROUPS   SERVICEACCOUNTS
rolebinding.rbac.authorization.k8s.io/read-pods   18s   ClusterRole/nginx-ingress-microk8s-clusterrole                    t1/build-robot
```

```bash
$ kubectl get statefulset mariadb-k8s -n t1 -o=jsonpath='{.spec.template.spec.serviceAccountName}'
build-robot
```

```bash
$ kubectl get sa build-robot -n t1 -o yaml
apiVersion: v1
automountServiceAccountToken: true
kind: ServiceAccount
metadata:
  creationTimestamp: "2019-07-05T08:44:10Z"
  labels:
    juju-app: mariadb-k8s
    juju-model: t1
  name: build-robot
  namespace: t1
  resourceVersion: "1099393"
  selfLink: /api/v1/namespaces/t1/serviceaccounts/build-robot
  uid: 1210cfec-00e2-44f5-973d-709948d1e77a
secrets:
- name: build-robot-token-s8px4
```

```bash
$ mkubectl get rolebinding.rbac.authorization.k8s.io/read-pods -n t1 -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  creationTimestamp: "2019-07-05T08:44:10Z"
  labels:
    juju-app: mariadb-k8s
    juju-model: t1
  name: read-pods
  namespace: t1
  resourceVersion: "1099392"
  selfLink: /apis/rbac.authorization.k8s.io/v1/namespaces/t1/rolebindings/read-pods
  uid: a91fc542-4e74-4d6e-b4fc-5601c8fd0b34
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: nginx-ingress-microk8s-clusterrole
subjects:
- kind: ServiceAccount
  name: build-robot
  namespace: t1
```


## Documentation changes

Required after RBAC supported.

## Bug reference

None
